crates/handlers/src/views/login.rs

```diff
@@ -146,181 +144,190 @@ pub(crate) async fn post(
+        )
+        .await;
+    };
 
-    let password = Zeroizing::new(password.as_bytes().to_vec());
+    let password = Zeroizing::new(form.password.as_bytes().to_vec());
 
     // Verify the password, and upgrade it on-the-fly if needed
-    let new_password_hash = password_manager
+    let user_password = match password_manager
         .verify_and_upgrade(
             &mut rng,
             user_password.version,
             password,
             user_password.hashed_password.clone(),
         )
         .await
-        .map_err(|_| FormError::InvalidCredentials)?;
-
-    let user_password = if let Some((version, new_password_hash)) = new_password_hash {
-        // Save the upgraded password
-        repo.user_password()
-            .add(
+    {
+        Ok(Some((version, new_password_hash))) => {
+            // Save the upgraded password
+            repo.user_password()
+                .add(
+                    &mut rng,
+                    &clock,
+                    &user,
+                    version,
+                    new_password_hash,
+                    Some(&user_password),
+                )
+                .await?
+        }
+        Ok(None) => user_password,
+        Err(_) => {
+            let form_state = form_state.with_error_on_form(FormError::InvalidCredentials);
+            return render(
+                locale,
+                cookie_jar,
+                form_state,
+                query,
+                &mut repo,
+                &clock,
                 &mut rng,
-                clock,
-                &user,
-                version,
-                new_password_hash,
-                Some(&user_password),
+                &templates,
+                &homeserver,
             )
-            .await
-            .map_err(|_| FormError::Internal)?
-    } else {
-        user_password
+            .await;
+        }
     };
 
+    // Now that we have checked the user password, we now want to show an error if
+    // the user is locked or deactivated
+    if user.deactivated_at.is_some() {
+        let (csrf_token, cookie_jar) = cookie_jar.csrf_token(&clock, &mut rng);
+        let ctx = AccountInactiveContext::new(user)
+            .with_csrf(csrf_token.form_value())
+            .with_language(locale);
+        let content = templates.render_account_deactivated(&ctx)?;
+        return Ok((cookie_jar, Html(content)).into_response());
+    }
+
+    if user.locked_at.is_some() {
+        let (csrf_token, cookie_jar) = cookie_jar.csrf_token(&clock, &mut rng);
+        let ctx = AccountInactiveContext::new(user)
+            .with_csrf(csrf_token.form_value())
+            .with_language(locale);
+        let content = templates.render_account_locked(&ctx)?;
+        return Ok((cookie_jar, Html(content)).into_response());
+    }
+
+    // At this point, we should have a 'valid' user. In case we missed something, we
+    // want it to crash in tests/debug builds
+    debug_assert!(user.is_valid());
```

Comment:
tbh, I'd be in favour of making this a full-blown assert.

I'd only use `debug_assert` for perf reasons, which doesn't seem to apply here.

Reply:
I get that, but I also kinda want tasks to not panic as much as possible? Alternative is to bubble up a proper Error but 🤷